worker/upload.go

```diff
@@ -0,0 +1,1152 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/montanaflynn/stats"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/mux/v1"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"go.sia.tech/renterd/tracing"
+	"lukechampine.com/frand"
+)
+
+var (
+	errNoFreeQueue    = errors.New("no free queue")
+	errNotEnoughHosts = errors.New("not enough hosts to support requested redundancy")
+)
+
+type (
+	uploadID [8]byte
+	slabID   [8]byte
+
+	uploadManager struct {
+		hp hostProvider
+		rl revisionLocker
+
+		maxOverdrive     uint64
+		overdriveTimeout time.Duration
+
+		statsOverdrive *dataPoints
+		statsSpeed     *dataPoints
+		stopChan       chan struct{}
+
+		mu        sync.Mutex
+		uploaders []*uploader
+		uploads   map[uploadID]*upload
+	}
+
+	uploader struct {
+		fcid       types.FileContractID
+		hk         types.PublicKey
+		siamuxAddr string
+
+		mu                  sync.Mutex
+		bh                  uint64
+		consecutiveFailures uint64
+		queueChan           chan struct{}
+		queue               []*uploadReq
+
+		statsSpeed *dataPoints
+		stopChan   chan struct{}
+	}
+
+	upload struct {
+		mgr *uploadManager
```

Comment:
Can we avoid having a reference back to the manager? Maybe by passing the overdrive settings directly to the upload instead of through the manager?

Ideally we'd have a clear hierarchy that looks something like uploadManager -> uploader -> upload -> slabUpload and none of those point to their parent to avoid blocking the garbage collection. Especially the manager since it sticks around forever.

This should also implicitly help with the overall architecture. e.g. an `upload` can launch a `slabUpload` and wait for it to finish. Then choose whether another upload is necessary.

As soon as you violate that hierarchy you tempt yourself to do stuff like have the child call methods on the parent potentially leading to harder to understand design or even deadlocks.

Reply:
I cleaned it up in a bunch of places. I kept the manager reference in both the `upload` and `slabUpload` because both need it to schedule shard uploads and other reasons like registering what uploader was used by what slab and so on.

I think GC-wise it is safe. I tried introducing an `uploaders` type that more or less acts like an uploader pool, but it felt like a completely unnecessary extra abstraction. Almost everything would move over from the manager to that type and the worst thing is that the pool would still need access to the manager's state or forward calls to the manager.

I think we can still add that afterwards if you want to but I feel the current version is good (enough) from the perspective of the various types and their naming and so forth. It's pretty polished at this point I think.